File: heap/benchmarks/sorting_benchmark.c
/*
 * Benchmark: Heapsort vs Mergesort vs Quicksort in C
 * Goal: Visualize the constant factor in O(n log n) without interpreter
 * overhead
 *
 * Compile: gcc -O2 -o sorting_benchmark sorting_benchmark.c -lm
 * Run:     ./sorting_benchmark
 */

#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// ============================================================================
// HEAPSORT (in-place)
// ============================================================================

static inline void swap(int *a, int *b) {
  int tmp = *a;
  *a = *b;
  *b = tmp;
}

static void max_heapify(int *A, int size, int i) {
  int greatest = i;
  int l = 2 * i + 1;
  int r = 2 * i + 2;

  if (l < size && A[l] > A[greatest])
    greatest = l;
  if (r < size && A[r] > A[greatest])
    greatest = r;

  if (greatest != i) {
    swap(&A[i], &A[greatest]);
    max_heapify(A, size, greatest);
  }
}

void heap_sort(int *A, int n) {
  if (n <= 1)
    return;

  // Build max-heap (bottom-up)
  for (int i = (n - 2) / 2; i >= 0; i--) {
    max_heapify(A, n, i);
  }

  // Extract elements one by one
  for (int heap_size = n - 1; heap_size > 0; heap_size--) {
    swap(&A[0], &A[heap_size]);
    max_heapify(A, heap_size, 0);
  }
}

// ============================================================================
// MERGESORT (with auxiliary array)
// ============================================================================

static int *aux = NULL; // Pre-allocated auxiliary array

static void merge(int *A, int lo, int mid, int hi) {
  // Copy to aux
  for (int k = lo; k <= hi; k++) {
    aux[k] = A[k];
  }

  int i = lo, j = mid + 1;
  for (int k = lo; k <= hi; k++) {
    if (i > mid) {
      A[k] = aux[j++];
    } else if (j > hi) {
      A[k] = aux[i++];
    } else if (aux[i] <= aux[j]) {
      A[k] = aux[i++];
    } else {
      A[k] = aux[j++];
    }
  }
}

static void mergesort_recursive(int *A, int lo, int hi) {
  if (lo >= hi)
    return;
  int mid = lo + (hi - lo) / 2;
  mergesort_recursive(A, lo, mid);
  mergesort_recursive(A, mid + 1, hi);
  merge(A, lo, mid, hi);
}

void merge_sort(int *A, int n) {
  if (n <= 1)
    return;
  aux = (int *)malloc(n * sizeof(int));
  mergesort_recursive(A, 0, n - 1);
  free(aux);
  aux = NULL;
}

// ============================================================================
// QUICKSORT (in-place, median-of-three pivot)
// ============================================================================

static int median_of_three(int *A, int lo, int hi) {
  int mid = lo + (hi - lo) / 2;
  if (A[lo] > A[mid])
    swap(&A[lo], &A[mid]);
  if (A[lo] > A[hi])
    swap(&A[lo], &A[hi]);
  if (A[mid] > A[hi])
    swap(&A[mid], &A[hi]);
  return mid;
}

static int partition(int *A, int lo, int hi) {
  int pivot_idx = median_of_three(A, lo, hi);
  swap(&A[pivot_idx], &A[hi]); // Move pivot to end
  int pivot = A[hi];

  int i = lo - 1;
  for (int j = lo; j < hi; j++) {
    if (A[j] <= pivot) {
      i++;
      swap(&A[i], &A[j]);
    }
  }
  swap(&A[i + 1], &A[hi]);
  return i + 1;
}

static void quicksort_recursive(int *A, int lo, int hi) {
  if (lo >= hi)
    return;
  int p = partition(A, lo, hi);
  quicksort_recursive(A, lo, p - 1);
  quicksort_recursive(A, p + 1, hi);
}

void quick_sort(int *A, int n) {
  if (n <= 1)
    return;
  quicksort_recursive(A, 0, n - 1);
}

// ============================================================================
// BENCHMARKING
// ============================================================================

typedef void (*SortFunc)(int *, int);

// High-resolution timer
double get_time_sec(void) {
  struct timespec ts;
  clock_gettime(CLOCK_MONOTONIC, &ts);
  return ts.tv_sec + ts.tv_nsec / 1e9;
}

// Generate random array
void fill_random(int *A, int n, unsigned int seed) {
  srand(seed);
  for (int i = 0; i < n; i++) {
    A[i] = rand() % (n * 10);
  }
}

// Verify array is sorted
int is_sorted(int *A, int n) {
  for (int i = 1; i < n; i++) {
    if (A[i] < A[i - 1])
      return 0;
  }
  return 1;
}

double benchmark_sort(SortFunc sort_func, int n, int trials,
                      unsigned int base_seed) {
  int *data = (int *)malloc(n * sizeof(int));
  double total_time = 0.0;

  for (int trial = 0; trial < trials; trial++) {
    fill_random(data, n, base_seed + trial);

    double start = get_time_sec();
    sort_func(data, n);
    double end = get_time_sec();

    total_time += (end - start);

    // Verify correctness
    if (!is_sorted(data, n)) {
      fprintf(stderr, "ERROR: Array not sorted for n=%d, trial=%d\n", n, trial);
    }
  }

  free(data);
  return total_time / trials;
}

double compute_constant(int n, double time_sec) {
  if (n <= 1)
    return 0.0;
  return time_sec / (n * log2((double)n));
}

int main(void) {
  int sizes[] = {100,   500,    1000,   5000,    10000,
                 50000, 100000, 500000, 1000000, 5000000};
  int num_sizes = sizeof(sizes) / sizeof(sizes[0]);
  int trials = 5;
  unsigned int seed = 42;

  printf("Benchmarking sorting algorithms (C implementation)\n");
  printf("Trials per size: %d\n", trials);
  printf("Compiler optimizations: -O2\n\n");

  // Header
  printf("====================================================================="
         "=============================\n");
  printf("%12s | %18s | %18s | %18s\n", "SIZE", "HEAPSORT", "MERGESORT",
         "QUICKSORT");
  printf("%12s | %8s %8s | %8s %8s | %8s %8s\n", "", "time(ms)", "const",
         "time(ms)", "const", "time(ms)", "const");
  printf("====================================================================="
         "=============================\n");

  double heap_consts[10], merge_consts[10], quick_consts[10];
  int large_count = 0;

  for (int i = 0; i < num_sizes; i++) {
    int n = sizes[i];

    double heap_t = benchmark_sort(heap_sort, n, trials, seed);
    double merge_t = benchmark_sort(merge_sort, n, trials, seed);
    double quick_t = benchmark_sort(quick_sort, n, trials, seed);

    double heap_c = compute_constant(n, heap_t) * 1e9; // nanoseconds
    double merge_c = compute_constant(n, merge_t) * 1e9;
    double quick_c = compute_constant(n, quick_t) * 1e9;

    printf("%12d | %8.3f %8.2f | %8.3f %8.2f | %8.3f %8.2f\n", n, heap_t * 1000,
           heap_c, merge_t * 1000, merge_c, quick_t * 1000, quick_c);

    // Collect constants for n >= 1000
    if (n >= 1000) {
      heap_consts[large_count] = heap_c;
      merge_consts[large_count] = merge_c;
      quick_consts[large_count] = quick_c;
      large_count++;
    }
  }

  printf("====================================================================="
         "=============================\n");
  printf("Note: 'const' = time / (n * log2(n)) in nanoseconds\n");
  printf(
      "      A stable constant across sizes confirms O(n log n) behavior\n\n");

  // Compute averages
  double avg_heap = 0, avg_merge = 0, avg_quick = 0;
  for (int i = 0; i < large_count; i++) {
    avg_heap += heap_consts[i];
    avg_merge += merge_consts[i];
    avg_quick += quick_consts[i];
  }
  avg_heap /= large_count;
  avg_merge /= large_count;
  avg_quick /= large_count;

  printf("Average constants (for n >= 1000):\n");
  printf("  Heapsort:  %.2f ns\n", avg_heap);
  printf("  Mergesort: %.2f ns\n", avg_merge);
  printf("  Quicksort: %.2f ns\n", avg_quick);

  double fastest = avg_quick;
  if (avg_heap < fastest)
    fastest = avg_heap;
  if (avg_merge < fastest)
    fastest = avg_merge;

  printf("\nRelative to fastest:\n");
  printf("  Heapsort:  %.2fx\n", avg_heap / fastest);
  printf("  Mergesort: %.2fx\n", avg_merge / fastest);
  printf("  Quicksort: %.2fx\n", avg_quick / fastest);

  return 0;
}

File: heap/benchmarks/sorting_benchmark.py
"""
Benchmark: Heapsort vs Mergesort vs Quicksort
Goal: Visualize the constant factor in O(n log n) by computing time / (n * log2(n))
"""

import random
import time
import math
from typing import Callable


# =======================
# HEAPSORT (in-place)
# =======================
def heapsort(A: list[int], n: int):
    if n <= 1:
        return

    def max_heapify(i: int, size: int):
        greatest = i
        l = 2 * i + 1
        r = 2 * i + 2
        if l < size and A[l] > A[greatest]:
            greatest = l
        if r < size and A[r] > A[greatest]:
            greatest = r
        if greatest != i:
            A[i], A[greatest] = A[greatest], A[i]
            max_heapify(greatest, size)

    # Build max-heap (bottom-up)
    for i in range((n - 2) // 2, -1, -1):
        max_heapify(i, n)

    # Extract elements one by one
    for heap_size in range(n - 1, 0, -1):
        A[0], A[heap_size] = A[heap_size], A[0]
        max_heapify(0, heap_size)


# =======================
# MERGESORT (uses auxiliary space, but same style)
# =======================
def mergesort(A: list[int], n: int):
    if n <= 1:
        return

    aux = [0] * n  # Pre-allocate auxiliary array once

    def merge(lo: int, mid: int, hi: int):
        # Copy to aux
        for k in range(lo, hi + 1):
            aux[k] = A[k]

        i, j = lo, mid + 1
        for k in range(lo, hi + 1):
            if i > mid:
                A[k] = aux[j]
                j += 1
            elif j > hi:
                A[k] = aux[i]
                i += 1
            elif aux[i] <= aux[j]:
                A[k] = aux[i]
                i += 1
            else:
                A[k] = aux[j]
                j += 1

    def sort(lo: int, hi: int):
        if lo >= hi:
            return
        mid = lo + (hi - lo) // 2
        sort(lo, mid)
        sort(mid + 1, hi)
        merge(lo, mid, hi)

    sort(0, n - 1)


# =======================
# QUICKSORT (in-place, median-of-three pivot)
# =======================
def quicksort(A: list[int], n: int):
    if n <= 1:
        return

    def median_of_three(lo: int, hi: int) -> int:
        mid = lo + (hi - lo) // 2
        # Sort lo, mid, hi and return mid as pivot index
        if A[lo] > A[mid]:
            A[lo], A[mid] = A[mid], A[lo]
        if A[lo] > A[hi]:
            A[lo], A[hi] = A[hi], A[lo]
        if A[mid] > A[hi]:
            A[mid], A[hi] = A[hi], A[mid]
        return mid

    def partition(lo: int, hi: int) -> int:
        # Use median-of-three for pivot selection
        pivot_idx = median_of_three(lo, hi)
        A[pivot_idx], A[hi] = A[hi], A[pivot_idx]  # Move pivot to end
        pivot = A[hi]

        i = lo - 1
        for j in range(lo, hi):
            if A[j] <= pivot:
                i += 1
                A[i], A[j] = A[j], A[i]
        A[i + 1], A[hi] = A[hi], A[i + 1]
        return i + 1

    def sort(lo: int, hi: int):
        if lo >= hi:
            return
        p = partition(lo, hi)
        sort(lo, p - 1)
        sort(p + 1, hi)

    sort(0, n - 1)


# =======================
# BENCHMARKING
# =======================
def benchmark_sort(
    sort_func: Callable[[list[int], int], None],
    sizes: list[int],
    trials: int = 5,
    seed: int = 42,
) -> dict[int, float]:
    """
    Benchmark a sorting function across different sizes.
    Returns dict mapping size -> average time in seconds.
    """
    results = {}

    for n in sizes:
        times = []
        for trial in range(trials):
            # Generate fresh random data each trial
            random.seed(seed + trial)
            data = [random.randint(0, n * 10) for _ in range(n)]

            # Time only the sort
            start = time.perf_counter()
            sort_func(data, n)
            end = time.perf_counter()

            times.append(end - start)

        results[n] = sum(times) / len(times)

    return results


def compute_constant(size: int, time_sec: float) -> float:
    """Compute the constant factor: time / (n * log2(n))"""
    if size <= 1:
        return 0.0
    return time_sec / (size * math.log2(size))


def print_results(results: dict[str, dict[int, float]], sizes: list[int]):
    """Pretty print the benchmark results with constant factors."""

    # Header
    print("\n" + "=" * 90)
    print(f"{'SIZE':>12} | {'HEAPSORT':>18} | {'MERGESORT':>18} | {'QUICKSORT':>18}")
    print(
        f"{'':>12} | {'time (ms)':>8} {'const':>8} | {'time (ms)':>8} {'const':>8} | {'time (ms)':>8} {'const':>8}"
    )
    print("=" * 90)

    for n in sizes:
        heap_t = results["heapsort"][n]
        merge_t = results["mergesort"][n]
        quick_t = results["quicksort"][n]

        heap_c = compute_constant(n, heap_t)
        merge_c = compute_constant(n, merge_t)
        quick_c = compute_constant(n, quick_t)

        # Time in milliseconds, constant in nanoseconds (more readable)
        print(
            f"{n:>12,} | {heap_t*1000:>8.3f} {heap_c*1e9:>8.2f} | "
            f"{merge_t*1000:>8.3f} {merge_c*1e9:>8.2f} | "
            f"{quick_t*1000:>8.3f} {quick_c*1e9:>8.2f}"
        )

    print("=" * 90)
    print("Note: 'const' = time / (n * log2(n)) in nanoseconds")
    print("      A stable constant across sizes confirms O(n log n) behavior")
    print()

    # Compute average constants (excluding smallest sizes)
    large_sizes = [n for n in sizes if n >= 1000]
    if large_sizes:
        avg_heap = (
            sum(compute_constant(n, results["heapsort"][n]) for n in large_sizes)
            / len(large_sizes)
            * 1e9
        )
        avg_merge = (
            sum(compute_constant(n, results["mergesort"][n]) for n in large_sizes)
            / len(large_sizes)
            * 1e9
        )
        avg_quick = (
            sum(compute_constant(n, results["quicksort"][n]) for n in large_sizes)
            / len(large_sizes)
            * 1e9
        )

        print("Average constants (for n >= 1000):")
        print(f"  Heapsort:  {avg_heap:.2f} ns")
        print(f"  Mergesort: {avg_merge:.2f} ns")
        print(f"  Quicksort: {avg_quick:.2f} ns")

        fastest = min(avg_heap, avg_merge, avg_quick)
        print(f"\nRelative to fastest:")
        print(f"  Heapsort:  {avg_heap/fastest:.2f}x")
        print(f"  Mergesort: {avg_merge/fastest:.2f}x")
        print(f"  Quicksort: {avg_quick/fastest:.2f}x")


def main():
    # Test sizes: from small to large
    sizes = [100, 500, 1_000, 5_000, 10_000, 50_000, 100_000, 500_000]

    # Increase recursion limit for large arrays
    import sys

    sys.setrecursionlimit(100_000)

    print("Benchmarking sorting algorithms...")
    print(f"Sizes: {sizes}")
    print(f"Trials per size: 5")
    print()

    # Run benchmarks
    print("Running heapsort...", end=" ", flush=True)
    heap_results = benchmark_sort(heapsort, sizes)
    print("done")

    print("Running mergesort...", end=" ", flush=True)
    merge_results = benchmark_sort(mergesort, sizes)
    print("done")

    print("Running quicksort...", end=" ", flush=True)
    quick_results = benchmark_sort(quicksort, sizes)
    print("done")

    results = {
        "heapsort": heap_results,
        "mergesort": merge_results,
        "quicksort": quick_results,
    }

    print_results(results, sizes)


if __name__ == "__main__":
    main()
